File: core/verdict.lua
-- core/verdict.lua
-- Flat verdict structure. Every field always present, no nested tables,
-- so it maps 1:1 onto JSON, protobuf and HTTP headers.

local _M = {}

_M.ACTION_PASS  = "pass"
_M.ACTION_BLOCK = "block"

_M.SAFE       = "safe"
_M.SUSPICIOUS = "suspicious"
_M.MALICIOUS  = "malicious"
_M.ERROR      = "error"
_M.SKIPPED    = "skipped"

_M.SRC_L1      = "l1"
_M.SRC_CACHE   = "cache"
_M.SRC_TRUST   = "trust"
_M.SRC_L2      = "l2"
_M.SRC_BREAKER = "breaker"

local function clamp01(n)
  if type(n) ~= "number" or n ~= n then return 0 end
  if n < 0 then return 0 end
  if n > 1 then return 1 end
  return n
end

--- Build a verdict with defaults filled in.
-- @param t table with any of: action, verdict, score, source, reason,
--          fingerprint, l2_ms, async, error_kind
-- error_kind: what an L2 error verdict ran into (judge.error_kind:
-- transport, timeout, unavailable, rejected, unusable, busy, other); "" on
-- every other verdict. It labels jev_l2_errors_total, so an operator can
-- tell a provider that is down from one that refuses every call.
function _M.new(t)
  t = t or {}
  return {
    action      = t.action or _M.ACTION_PASS,
    verdict     = t.verdict or _M.SKIPPED,
    score       = clamp01(t.score or 0),
    source      = t.source or _M.SRC_L1,
    reason      = tostring(t.reason or ""),
    fingerprint = tostring(t.fingerprint or ""),
    l2_ms       = tonumber(t.l2_ms) or 0,
    async       = t.async == true,
    error_kind  = tostring(t.error_kind or ""),
  }
end

--- Headers to set on the upstream request.
function _M.headers(v)
  return {
    ["X-Jev-Verdict"] = v.verdict,
    ["X-Jev-Score"]   = string.format("%.2f", v.score),
    ["X-Jev-Source"]  = v.source,
    ["X-Jev-Reason"]  = _M.encode_reason(v.reason),
  }
end

--- The only verdict header a client may see, on a block response (with
--- X-Jev-Request-Id, which the adapter adds). Score, reason and source say
--- how close a text came and which pattern or template fired: an attacker
--- tunes a prompt against them, so they go to the logs and the upstream
--- request (headers above), never back to the client.
function _M.client_headers(v)
  return { ["X-Jev-Verdict"] = v.verdict }
end

--- URL-encode and truncate a reason for header transport (<= 200 bytes of
--- encoded output, never cut inside a %XX escape).
_M.REASON_MAX = 200
function _M.encode_reason(s)
  s = tostring(s or "")
  local enc = s:gsub("[^%w%-%._~ ]", function(c)
    return string.format("%%%02X", string.byte(c))
  end):gsub(" ", "+")
  if #enc > _M.REASON_MAX then
    enc = enc:sub(1, _M.REASON_MAX):gsub("%%%x?$", "")
  end
  return enc
end

return _M

File: adapters/js/src/core/verdict.ts
// Port of core/verdict.lua. Flat structure: every field always present.
import { utf8Bytes } from "./normalize.js";

export type Action = "pass" | "block";
export type Label = "safe" | "suspicious" | "malicious" | "error" | "skipped";
/** "adapter" is what the host sets when its own pipeline threw and the request failed open (edge.lua access()). */
export type Source = "l1" | "cache" | "trust" | "l2" | "breaker" | "adapter";

export const ACTION_PASS: Action = "pass";
export const ACTION_BLOCK: Action = "block";
export const SAFE: Label = "safe";
export const SUSPICIOUS: Label = "suspicious";
export const MALICIOUS: Label = "malicious";
export const ERROR: Label = "error";
export const SKIPPED: Label = "skipped";
export const SRC_L1: Source = "l1";
export const SRC_CACHE: Source = "cache";
export const SRC_TRUST: Source = "trust";
export const SRC_L2: Source = "l2";
export const SRC_BREAKER: Source = "breaker";
export const SRC_ADAPTER: Source = "adapter";

export interface Verdict {
  action: Action;
  verdict: Label;
  score: number;
  source: Source;
  reason: string;
  fingerprint: string;
  l2_ms: number;
  async: boolean;
}

export interface VerdictInit {
  action?: Action;
  verdict?: Label;
  score?: number;
  source?: Source;
  reason?: unknown;
  fingerprint?: unknown;
  l2_ms?: unknown;
  async?: unknown;
}

function clamp01(n: unknown): number {
  if (typeof n !== "number" || Number.isNaN(n)) return 0;
  if (n < 0) return 0;
  if (n > 1) return 1;
  return n;
}

export function newVerdict(t?: VerdictInit | null): Verdict {
  const v = t ?? {};
  const l2 = typeof v.l2_ms === "number" ? v.l2_ms : Number(v.l2_ms);
  return {
    action: v.action ?? ACTION_PASS,
    verdict: v.verdict ?? SKIPPED,
    score: clamp01(v.score ?? 0),
    source: v.source ?? SRC_L1,
    reason: String(v.reason ?? ""),
    fingerprint: String(v.fingerprint ?? ""),
    l2_ms: Number.isFinite(l2) ? l2 : 0,
    async: v.async === true,
  };
}

/**
 * C printf("%.2f"): exact binary value, ties to even. JS toFixed rounds ties
 * away from zero on the *decimal* expansion, which differs for e.g. 0.125.
 */
export function format2(n: number): string {
  const exact = n.toFixed(20); // exact enough: doubles in [0,1] need <= 17 significant digits
  const [ip, fp = ""] = exact.split(".");
  const keep = fp.slice(0, 2).padEnd(2, "0");
  const rest = fp.slice(2);
  let carry = 0;
  if (rest.length > 0) {
    const first = rest.charCodeAt(0) - 48;
    const tail = rest.slice(1);
    if (first > 5 || (first === 5 && /[1-9]/.test(tail))) carry = 1;
    else if (first === 5) carry = (keep.charCodeAt(1) - 48) % 2 === 1 ? 1 : 0; // tie: to even
  }
  let cents = parseInt(ip, 10) * 100 + parseInt(keep, 10) + carry;
  const neg = n < 0 && cents !== 0;
  cents = Math.abs(cents);
  const s = `${Math.floor(cents / 100)}.${String(cents % 100).padStart(2, "0")}`;
  return neg ? "-" + s : s;
}

/** Longest encoded reason placed in a header, in bytes. */
export const REASON_MAX = 200;

/** URL-encode and truncate a reason for header transport (<= REASON_MAX bytes of encoded output, never cut inside a %XX escape). */
export function encodeReason(s: unknown): string {
  const str = String(s ?? "");
  // Lua: gsub("[^%w%-%._~ ]", %%XX) over bytes, then spaces to +
  let out = "";
  for (const b of utf8Bytes(str)) {
    const c = String.fromCharCode(b);
    if (/[A-Za-z0-9\-._~ ]/.test(c) && b < 128) out += c === " " ? "+" : c;
    else out += "%" + b.toString(16).toUpperCase().padStart(2, "0");
  }
  // every char of `out` is one ASCII byte, so length is bytes
  if (out.length > REASON_MAX) out = out.slice(0, REASON_MAX).replace(/%[0-9A-Fa-f]?$/, "");
  return out;
}

/** Port of verdict.client_headers: the only verdict header a client may see
 *  on a block response (with X-Jev-Request-Id, which the adapter adds).
 *  Score, reason and source go to the logs and the upstream request only. */
export function clientHeaders(v: Verdict): Record<string, string> {
  return { "X-Jev-Verdict": v.verdict };
}

export function headers(v: Verdict): Record<string, string> {
  return {
    "X-Jev-Verdict": v.verdict,
    "X-Jev-Score": format2(v.score),
    "X-Jev-Source": v.source,
    "X-Jev-Reason": encodeReason(v.reason),
  };
}
